// kwbl-gen/generator.js
var path = require('path');
//var util = require('util');

var generators = {
	uuid : genGUID,
	guid : genGUID,
	uid: uid,
	couchTimeStamp : getTime,
	scriptName : getScriptName,
	cost : genCost,
	costToStr : genCostStr,
	duration : genDuration,
	random : genRandom
}

function genRandom(min, max){
	if(min instanceof Array){
		return randomArrayElement(min, max);
	}

	if(typeof min === 'object'){
		return randomObjectElement(min, max);
	}

	if(typeof max === 'undefined'){
		max = min;
		min = undefined;
	}

	if(typeof min === 'undefined'){
		min = 0;
	}

	if(typeof max === 'undefined'){
		max = 10;
	}

	

	if(min > max){
		var tmp = min;
		min = max;
		max = tmp;
	}

	var diff = 0;

	if(min < 0){
		var diff = - min;
		min = min + diff;
		max = max + diff;
	}

	var r = Math.random() * (max+1-min)
	//console.log(min, max, r, (min + r));

	var res = Math.floor(r) + min;

	return diff === 0 ? res : res - diff;
}

function randomArrayElement(arr, keyOnly){
	var key = genRandom(arr.length-1);
	if(keyOnly === true || keyOnly === 'key'){
		return key;
	}
	//console.log('randomArrayElement', key, arr);
	return arr[key];
}

function randomObjectElement(obj, keyOnly){
	var key = randomArrayElement(Object.keys(obj));
	if(keyOnly === true || keyOnly === 'key'){
		return key;
	}
	//console.log('randomObjectElement', key);
	return obj[key];
}

function genDuration(){
	return genRandom(30*60);
}

function genCost(mult){
	if(! mult) mult = 1;
	//??? genRandom(100000)
	return mult*genRandom(Math.pow(10, genRandom(1, 5))); // /100000
}

function genCostStr(cost){
	if(typeof cost === 'undefined'){
		cost = genCost();
	}

	var costNum = cost/100000;

	var costInt = Math.floor(costNum);

	var costStr = costNum + "";

	var costlength = (costInt + "").length + '.00000'.length; 

	if(costStr.length === costlength){
		return costStr;
	}

	return costStr + (new Array(costlength - costStr.length + 1 )).join('0');

}

function getScriptName(filepath){
	if(! filepath) filepath = __filename;

  var start = filepath.lastIndexOf(path.sep)+1;
  var end = filepath.lastIndexOf('.');

  return filepath.substring(start,end);
}

function getTime(shiftOrDay, shift){

	var date = null;

	if(shiftOrDay instanceof Date){
		date = shiftOrDay;
	} else if(! shift && ! isNaN(shiftOrDay*1)){
		shift = shiftOrDay;
	}

	if(! date){
		date = new Date();
	}

  if(shift){
  	date = new Date(date.getTime() + shift);
  }

  return {
    t: date.getTime(), 
    Y: date.getFullYear(), 
    M: date.getMonth()+1, 
    D: date.getDate(), 
    h: date.getHours(), 
    m: date.getMinutes(), 
    s: date.getSeconds(), 
    z: "UTC" + (date.getTimezoneOffset() < 0? "+" + (-date.getTimezoneOffset()/60) : "-" + date.getTimezoneOffset()/60)
  };
}


function genGUID () {
  function s4() {
    return Math.floor((1 + Math.random()) * 0x10000)
               .toString(16)
               .substring(1);
  }
  return s4() + s4() + '-' + s4() + '-' + s4() + '-' +
           s4() + '-' + s4() + s4() + s4();
};

function uid (len) {
  var buf = [];
  var chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  var charlen = chars.length;

  len = len || 8;

  for (var i = 0; i < len; ++i) {
    buf.push(chars[genRandom(charlen - 1)]);
  }

  return buf.join('');
};

module.exports = generators;
